File: Source/WebKit/Scripts/webkit/tests/TestWithSwiftMessageReceiver.cpp
/*
 * Copyright (C) 2021-2023 Apple Inc. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1.  Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 * 2.  Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "config.h"
#include "Shared/WebKit-Swift.h" // NOLINT
#include "Decoder.h" // NOLINT
#include "HandleMessage.h" // NOLINT
#include "TestWithSwiftMessages.h" // NOLINT

#if ENABLE(IPC_TESTING_API)
#include "JSIPCBinding.h"
#endif

namespace WebKit {

void TestWithSwiftMessageForwarder::didReceiveMessage(IPC::Connection& connection, IPC::Decoder& decoder)
{
    Ref protectedThis { *this };
    auto target = getMessageTarget();
    if (!target) {
        FATAL("Something is keeping a reference to the message forwarder");
        decoder.markInvalid();
        return;
    }
    if (decoder.messageName() == Messages::TestWithSwift::TestAsyncMessage::name()) {
        IPC::handleMessageAsync<Messages::TestWithSwift::TestAsyncMessage>(connection, decoder, target.get(), &TestWithSwift::testAsyncMessage);
        return;
    }
    UNUSED_PARAM(connection);
    RELEASE_LOG_ERROR(IPC, "Unhandled message %s to %" PRIu64, IPC::description(decoder.messageName()).characters(), decoder.destinationID());
    decoder.markInvalid();
}

void TestWithSwiftMessageForwarder::didReceiveSyncMessage(IPC::Connection& connection, IPC::Decoder& decoder, UniqueRef<IPC::Encoder>& replyEncoder)
{
    Ref protectedThis { *this };
    auto target = getMessageTarget();
    if (!target) {
        FATAL("Something is keeping a reference to the message forwarder");
        decoder.markInvalid();
        return;
    }
    if (decoder.messageName() == Messages::TestWithSwift::TestSyncMessage::name()) {
        IPC::handleMessageSynchronous<Messages::TestWithSwift::TestSyncMessage>(connection, decoder, replyEncoder, target.get(), &TestWithSwift::testSyncMessage);
        return;
    }
    UNUSED_PARAM(connection);
    UNUSED_PARAM(replyEncoder);
    RELEASE_LOG_ERROR(IPC, "Unhandled synchronous message %s to %" PRIu64, description(decoder.messageName()).characters(), decoder.destinationID());
    decoder.markInvalid();
}

static std::unique_ptr<TestWithSwiftWeakRef> makeTestWithSwiftWeakRefUniquePtr(TestWithSwiftWeakRef* _Nonnull handler)
{
    auto newRef = _impl::_impl_TestWithSwiftWeakRef::makeRetained(handler);
    return WTF::makeUniqueWithoutFastMallocCheck<TestWithSwiftWeakRef>(newRef);
}

TestWithSwiftMessageForwarder::TestWithSwiftMessageForwarder(TestWithSwiftWeakRef* _Nonnull target)
    : m_handler(makeTestWithSwiftWeakRefUniquePtr(target))
{
}

std::unique_ptr<TestWithSwift> TestWithSwiftMessageForwarder::getMessageTarget()
{
    auto target = m_handler->getMessageTarget();
    if (target)
        return WTF::makeUniqueWithoutFastMallocCheck<TestWithSwift>(target.get());
    return nullptr;
}

TestWithSwiftMessageForwarder::~TestWithSwiftMessageForwarder()
{
}


} // namespace WebKit

#if ENABLE(IPC_TESTING_API)

namespace IPC {

template<> std::optional<JSC::JSValue> jsValueForDecodedMessage<MessageName::TestWithSwift_TestAsyncMessage>(JSC::JSGlobalObject* globalObject, Decoder& decoder)
{
    return jsValueForDecodedArguments<Messages::TestWithSwift::TestAsyncMessage::Arguments>(globalObject, decoder);
}
template<> std::optional<JSC::JSValue> jsValueForDecodedMessageReply<MessageName::TestWithSwift_TestAsyncMessage>(JSC::JSGlobalObject* globalObject, Decoder& decoder)
{
    return jsValueForDecodedArguments<Messages::TestWithSwift::TestAsyncMessage::ReplyArguments>(globalObject, decoder);
}
template<> std::optional<JSC::JSValue> jsValueForDecodedMessage<MessageName::TestWithSwift_TestSyncMessage>(JSC::JSGlobalObject* globalObject, Decoder& decoder)
{
    return jsValueForDecodedArguments<Messages::TestWithSwift::TestSyncMessage::Arguments>(globalObject, decoder);
}
template<> std::optional<JSC::JSValue> jsValueForDecodedMessageReply<MessageName::TestWithSwift_TestSyncMessage>(JSC::JSGlobalObject* globalObject, Decoder& decoder)
{
    return jsValueForDecodedArguments<Messages::TestWithSwift::TestSyncMessage::ReplyArguments>(globalObject, decoder);
}
template<> std::optional<JSC::JSValue> jsValueForDecodedMessage<MessageName::TestWithSwift_TestAsyncMessageReply>(JSC::JSGlobalObject* globalObject, Decoder& decoder)
{
    return jsValueForDecodedArguments<Messages::TestWithSwift::TestAsyncMessageReply::Arguments>(globalObject, decoder);
}

}

#endif

File: Source/WebKit/Scripts/webkit/tests/TestWithSwiftMessageReceiver.swift
//
// Copyright (C) 2021-2023 Apple Inc. All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions
// are met:
// 1.  Redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
// 2.  Redistributions in binary form must reproduce the above copyright
// notice, this list of conditions and the following disclaimer in the
// documentation and/or other materials provided with the distribution.
//
// THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS'' AND
// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
// DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS BE LIABLE FOR
// ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//

import WebKit_Internal

final class TestWithSwiftWeakRef: @unchecked Sendable {
    private weak var target: TestWithSwift?
    init(target: TestWithSwift) {
        self.target = target
    }

    @used
    func getMessageTarget() -> TestWithSwift? {
        target
    }

    @used
    func dispatchTestAsyncMessage(
        connection: sending IPC.Connection,
        param: sending UInt32,
        completionHandler: sending CompletionHandlers.TestWithSwift.TestAsyncMessageCompletionHandler
    ) {
        MainActor.assumeIsolated {
            guard let target else {
                return
            }
            Task.immediate {
                await Self.runTestAsyncMessage(
                    target: target,
                    connection: connection,
                    param: param,
                    completionHandler: completionHandler
                )
            }
        }
    }

    @MainActor
    private static func runTestAsyncMessage(
        target: TestWithSwift,
        connection: IPC.Connection,
        param: UInt32,
        completionHandler: CompletionHandlers.TestWithSwift.TestAsyncMessageCompletionHandler
    ) async {
        do {
            let reply = try await mayThrowInvalidMessage(
                target.testAsyncMessage(
                    connection: connection,
                    param: param
                )
            )
            completionHandler.pointee(reply)
        } catch {
            markMessageInvalid(error, on: connection, message: .TestWithSwift_TestAsyncMessage)
            CompletionHandlers.TestWithSwift.completeWithDefaultReply(completionHandler)
        }
    }

    @used
    func dispatchTestSyncMessage(
        connection: sending IPC.Connection,
        param: sending UInt32,
        completionHandler: sending CompletionHandlers.TestWithSwift.TestSyncMessageCompletionHandler
    ) {
        MainActor.assumeIsolated {
            guard let target else {
                return
            }
            Task.immediate {
                await Self.runTestSyncMessage(
                    target: target,
                    connection: connection,
                    param: param,
                    completionHandler: completionHandler
                )
            }
        }
    }

    @MainActor
    private static func runTestSyncMessage(
        target: TestWithSwift,
        connection: IPC.Connection,
        param: UInt32,
        completionHandler: CompletionHandlers.TestWithSwift.TestSyncMessageCompletionHandler
    ) async {
        do {
            let reply = try await mayThrowInvalidMessage(
                target.testSyncMessage(
                    connection: connection,
                    param: param
                )
            )
            completionHandler.pointee(reply)
        } catch {
            markMessageInvalid(error, on: connection, message: .TestWithSwift_TestSyncMessage)
            CompletionHandlers.TestWithSwift.completeWithDefaultReply(completionHandler)
        }
    }

    @used
    func dispatchTestMessageWithAliasedParameter(
        connection: sending IPC.Connection,
        frameState: sending WebKit.RefFrameState
    ) {
        MainActor.assumeIsolated {
            guard let target else {
                return
            }
            Self.runTestMessageWithAliasedParameter(
                target: target,
                connection: connection,
                frameState: frameState
            )
        }
    }

    @MainActor
    private static func runTestMessageWithAliasedParameter(
        target: TestWithSwift,
        connection: IPC.Connection,
        frameState: WebKit.RefFrameState
    ) {
        do {
            try mayThrowInvalidMessage(
                target.testMessageWithAliasedParameter(
                    connection: connection,
                    frameState: frameState
                )
            )
        } catch {
            markMessageInvalid(error, on: connection, message: .TestWithSwift_TestMessageWithAliasedParameter)
        }
    }

    @used
    func dispatchTestThrowingMessageWithReply(
        connection: sending IPC.Connection,
        param: sending UInt32,
        completionHandler: sending CompletionHandlers.TestWithSwift.TestThrowingMessageWithReplyCompletionHandler
    ) {
        MainActor.assumeIsolated {
            guard let target else {
                return
            }
            Task.immediate {
                await Self.runTestThrowingMessageWithReply(
                    target: target,
                    connection: connection,
                    param: param,
                    completionHandler: completionHandler
                )
            }
        }
    }

    @MainActor
    private static func runTestThrowingMessageWithReply(
        target: TestWithSwift,
        connection: IPC.Connection,
        param: UInt32,
        completionHandler: CompletionHandlers.TestWithSwift.TestThrowingMessageWithReplyCompletionHandler
    ) async {
        do {
            let reply = try await mayThrowInvalidMessage(
                target.testThrowingMessageWithReply(
                    connection: connection,
                    param: param
                )
            )
            completionHandler.pointee(reply)
        } catch {
            markMessageInvalid(error, on: connection, message: .TestWithSwift_TestThrowingMessageWithReply)
            CompletionHandlers.TestWithSwift.completeWithDefaultReply(completionHandler)
        }
    }

    @used
    func dispatchTestThrowingMessageWithoutReply(
        connection: sending IPC.Connection,
        frameState: sending WebKit.RefFrameState,
        frameID: sending WebCore.FrameIdentifier
    ) {
        MainActor.assumeIsolated {
            guard let target else {
                return
            }
            Self.runTestThrowingMessageWithoutReply(
                target: target,
                connection: connection,
                frameState: frameState,
                frameID: frameID
            )
        }
    }

    @MainActor
    private static func runTestThrowingMessageWithoutReply(
        target: TestWithSwift,
        connection: IPC.Connection,
        frameState: WebKit.RefFrameState,
        frameID: WebCore.FrameIdentifier
    ) {
        do {
            try mayThrowInvalidMessage(
                target.testThrowingMessageWithoutReply(
                    connection: connection,
                    frameState: frameState,
                    frameID: frameID
                )
            )
        } catch {
            markMessageInvalid(error, on: connection, message: .TestWithSwift_TestThrowingMessageWithoutReply)
        }
    }
}

extension WebKit.TestWithSwiftMessageForwarder {
    static func create(target: TestWithSwift) -> RefTestWithSwiftMessageForwarder {
        let weakRefContainer = TestWithSwiftWeakRef(target: target)
        // Safety: we're creating a pointer which will immediately be stored in a
        // proper ref-counted reference on the C++ side before this call returns.
        // Workaround for rdar://163107752.
        return unsafe WebKit.TestWithSwiftMessageForwarder.createFromWeak(
            OpaquePointer(
                Unmanaged.passRetained(weakRefContainer).toOpaque()
            )
        )
    }
}
